core/src/main/java/org/elasticsearch/index/percolator/PercolatorFieldMapper.java

```diff
@@ -152,8 +174,37 @@ public PercolatorFieldMapper(String simpleName, MappedFieldType fieldType, Mappe
     }
 
     @Override
+    public FieldMapper updateFieldType(Map<String, MappedFieldType> fullNameToFieldType) {
+        PercolatorFieldMapper updated = (PercolatorFieldMapper) super.updateFieldType(fullNameToFieldType);
+        KeywordFieldMapper queryTermsUpdated = (KeywordFieldMapper) queryTermsField.updateFieldType(fullNameToFieldType);
+        KeywordFieldMapper unknownQueryUpdated = (KeywordFieldMapper) unknownQueryField.updateFieldType(fullNameToFieldType);
+        BinaryFieldMapper queryBuilderUpdated = (BinaryFieldMapper) queryBuilderField.updateFieldType(fullNameToFieldType);
+        if (updated == this && queryTermsUpdated == queryTermsField && unknownQueryUpdated == unknownQueryField &&
+                queryBuilderUpdated == queryBuilderField) {
+            return this;
+        }
+        if (updated == this) {
+            updated = (PercolatorFieldMapper) updated.clone();
+        }
+        updated.queryTermsField = queryTermsUpdated;
+        updated.unknownQueryField = unknownQueryUpdated;
+        updated.queryBuilderField = queryBuilderUpdated;
+        return updated;
+    }
+
+    @Override
     public Mapper parse(ParseContext context) throws IOException {
         QueryShardContext queryShardContext = new QueryShardContext(this.queryShardContext);
+        DocumentMapper documentMapper = queryShardContext.getMapperService().documentMapper(context.type());
+        for (FieldMapper fieldMapper : documentMapper.mappers()) {
+            if (fieldMapper instanceof PercolatorFieldMapper) {
+                PercolatorFieldType fieldType = (PercolatorFieldType) fieldMapper.fieldType();
+                if (context.doc().getField(fieldType.getQueryBuilderFieldName()) != null) {
+                    throw new IllegalArgumentException("a document can only contain one percolator query");
+                }
+            }
+        }
```

Comment:
In the case that the percolator field is part of another object field and during indexing this object field gets specified as array then we need to restrict that multiple queries are specified inside a single document. (there is test in `PercolatorFieldMapperTests` that covers this)